pubsub/iterator.go

```diff
@@ -528,39 +518,103 @@ func (it *messageIterator) sendModAck(m map[string]*ackResultWithID, deadline ti
 		exactlyOnceDelivery := it.enableExactlyOnceDelivery
 		it.eoMu.RUnlock()
 		if exactlyOnceDelivery {
-			var md map[string]string
-			var st *status.Status
-			apiErr, ok := apierror.FromError(err)
-			if ok {
-				md = apiErr.Metadata()
-				st = apiErr.GRPCStatus()
-			}
-			resultsByAckID := make(map[string]*ackResultWithID)
+			resultsByAckID := make(map[string]*AckResult)
 			for _, ackID := range toSend {
 				resultsByAckID[ackID] = m[ackID]
 			}
-			// TODO(hongalex): retry the ackIDs with transient retriable errors.
-			_, _ = processResults(st, resultsByAckID, md)
+
+			st, md := extractMetadata(err)
+			_, toRetry := processResults(st, resultsByAckID, md)
+			if len(toRetry) > 0 {
+				// Retry modacks/nacks in a separate goroutine.
+				go func() {
+					it.retryModAcks(toRetry, deadlineSec)
+				}()
+			}
 		}
 	}
 }
 
-func (it *messageIterator) sendAckIDRPC(ackResultsByAckID map[string]*AckResult, maxSize int, call func([]string) error) bool {
-	ackIDs := make([]string, 0, len(ackResultsByAckID))
-	for k := range ackResultsByAckID {
-		ackIDs = append(ackIDs, k)
+func (it *messageIterator) retryAcks(m map[string]*AckResult) {
+	ctx, cancel := context.WithTimeout(context.Background(), exactlyOnceDeliveryRetryDeadline)
+	defer cancel()
+	bo := newExactlyOnceBackoff()
+	for {
```

Comment:
I don't think we do this in the Java client library, but thoughts on adding a similar retry count that exists in the `retryModAcks` here?

Reply:
I can see adding this retry count for metrics but it's out of the scope of Prad's PR sequencing doc I think. IIRC, we cap modacks retries since after a while messages will just be expired.